composer/trainer/trainer.py

```diff
@@ -459,19 +524,84 @@ def _get_metrics_as_collection(self, *, is_train: bool) -> MetricCollection:
 
         return metrics
 
-    def _compute_and_log_metrics(self, metrics: Metrics, *, is_train: bool, is_batch: bool):
+    def _create_evaluator_metrics(self, evaluator: Evaluator = None) -> MetricCollection:
+        """Get metrics compatible with a model and deepcopy them to store in an Evaluator.
+        
+        Get metrics relevant to a model or a particular evaluator. Metrics
+        are all implemented as subclasses of :class:`torchmetrics.Metric`. This
+        function returns metrics as a :class:`~torchmetrics.collections.MetricCollection`
+        to enable support for multiple metrics.
+
+        Args:
+            evaluator (Evaluator): Evaluator to get the relevant metrics for. If the Evaluator
+                metric_names is empty or None is provided, the function returns
+                a copy of all the model's default evaluation metrics.
+
+        Returns:
+            A :class:`~torchmetrics.collections.MetricCollection` object.
+        """
+        # Get and copy all the model's associated metrics
+        original_model = self.state.model
+        assert isinstance(original_model, BaseMosaicModel)
+        metrics = original_model.metrics(train=False)
+        assert isinstance(metrics, (Metric, MetricCollection)), \
+        "   Error module.metrics() must return a Metric or MetricCollection object."
+        if isinstance(metrics, Metric):
+            # Forcing metrics to be a MetricCollection simplifies logging results
+            metrics = MetricCollection([metrics])
+
+        # Use all the metrics from the model if no metric_names are specified
+        if evaluator.metric_names is None or evaluator.metrics is None or evaluator.metrics:
```

Comment:
This is more of a design question, but should metrics be specified on the model or in the evaluators? I'd be for having just one place for them -- ideally the evaulators calss? I think it can be confusing if they could be specified in multiple places. Would it be ok to remove the `metrics` from the `BaseMosaicModel` and instead require that metrics be specified as `Evaluator`s?

Reply:
One issue with that suggestion is that right now we have the requirement that the output of a model's `validate()` can be passed directly into the `update()` function for all metrics that are computed for the model.

If the specification of the `metrics` is completely disjoint from the model (ex. by specifying metrics only in `Evaluator`s), then it opens up the possibility of bugs where the output of a model's `validate()` is not compatible with a metric specified in an `Evaluator`.

Reply:
@ravi-mosaicml To add to my comment above:

What makes this tricky is that we want to be able to specify `Evaluator` objects in YAML files, which means that we can only specify `metrics` for the `Evaluator` using strings.

Metrics can either come from the `torchmetrics` library or be a user-defined class that extends `torchmetrics.Metrics`. Also, in the latter case, there may be arbitrary parameters needed to instantiate the metric.

Because there doesn't seem to be a clear way for an Evaluator to instantiate a `metric` object just from the name (and possibly parameter list), the cleanest solution seemed to be to have the model return the `metric` objects that are associated with it and then have each individual `Evaluator` select the relevant `metric` objects from what gets returned by `model.metrics()` using the name of the metric. 

Definitely open to alternative solutions if you have something in mind - these were just my thoughts.